File: Leetcode/C++/200. Number of Islands/solution.cpp
// https://leetcode.com/problems/number-of-islands/

#include <vector>

using namespace std;

class Solution // DFS
{
public:
    int numIslands(vector<vector<char>> &grid)
    {
        int dx[4] = {0, 0, -1, 1}, dy[4] = {1, -1, 0, 0};
        int count = 0;
        for (int y = 0; y < grid.size(); y++)
        {
            for (int x = 0; x < grid[0].size(); x++)
            {
                if (grid[y][x] == '1')
                {
                    count += 1; // Count island
                    vector<pair<int, int>> stack;
                    stack.push_back({x, y});
                    grid[y][x] = '0'; // Check visit start point
                    while (!stack.empty())
                    {
                        int curX = stack.back().first, curY = stack.back().second;
                        stack.pop_back();
                        for (int idx = 0; idx < 4; idx++)
                        {
                            int newX = curX + dx[idx], newY = curY + dy[idx];
                            if (0 <= newX && newX < grid[0].size() && 0 <= newY && newY < grid.size() && grid[newY][newX] == '1')
                            {
                                grid[newY][newX] = '0';
                                stack.push_back({newX, newY});
                            }
                        }
                    }
                }
            }
        }

        return count;
    }
};

File: Leetcode/Typescript/200. Number of Islands/solution.ts
// https://leetcode.com/problems/number-of-islands/description/

interface point {
  x: number;
  y: number;
}

function numIslands(grid: string[][]): number {
  const stack: point[] = [];
  const dy = [0, 0, 1, -1],
    dx = [1, -1, 0, 0];
  const rowLength = grid.length,
    colLength = grid[0].length;
  let number = 0;

  for (let y = 0; y < rowLength; y++) {
    for (let x = 0; x < colLength; x++) {
      if (grid[y][x] === "1") {
        number += 1;
        stack.push({ x, y });

        while (stack.length > 0) {
          let cur = stack.pop();

          for (let i = 0; i < 4; i++) {
            const newX = cur!.x + dx[i],
              newY = cur!.y + dy[i];

            if (
              0 <= newX &&
              newX < colLength &&
              0 <= newY &&
              newY < rowLength &&
              grid[newY][newX] === "1"
            ) {
              grid[newY][newX] = "0";
              stack.push({ x: newX, y: newY });
            }
          }
        }
      }
    }
  }

  return number;
}

/**
 * Recursion
 */
const dy = [0, 0, 1, -1],
  dx = [1, -1, 0, 0];

function dfs(grid: string[][], x: number, y: number) {
  if (
    x < 0 ||
    grid[0].length <= x ||
    y < 0 ||
    grid.length <= y ||
    grid[y][x] === "0"
  ) {
    return;
  }
  grid[y][x] = "0";
  for (let idx = 0; idx < 4; idx++) {
    dfs(grid, x + dx[idx], y + dy[idx]);
  }
}
function numIslands(grid: string[][]): number {
  const rowLength = grid.length,
    colLength = grid[0].length;
  let number = 0;

  for (let y = 0; y < rowLength; y++) {
    for (let x = 0; x < colLength; x++) {
      if (grid[y][x] === "1") {
        dfs(grid, x, y);
        number += 1;
      }
    }
  }

  return number;
}
